<a href="https://colab.research.google.com/github/youngchanseo/Hon.Gong.Machine-SNU/blob/main/(05_3_2)_%EA%B7%B8%EB%9E%98%EB%94%94%EC%96%B8%ED%8A%B8_%EB%B6%80%EC%8A%A4%ED%8C%85_p271.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
wine = pd.read_csv('https://raw.githubusercontent.com/youngchanseo/Hon.Gong.Machine-SNU/main/(05_3_0)%20Raw%20data.csv')# https://bit.ly/wine_csv_data
data = wine[['alcohol', 'sugar', 'pH']].to_numpy()
target = wine['class'].to_numpy()
train_input, test_input, train_target, test_target = train_test_split(data, target, test_size=0.2, random_state=42)

# Gradient Boosting
 깊이가 얕은 결정 트리를 사용, 이전 트리의 오차를 보완하는 Ensemble

GradientBoostingClassifier
- 경사 하강법 사용하여 tree를 ensemble에 추가
- 기본적은로 깊이가 3인 결정 트리를 100개 사용
- 깊이가 얕은 결정 트리를 사용하여 과대 적합에 강함, 높은 일반화 성능


In [5]:
#GradientBoostingClassifier를 사용해 와인 데이터 셋의 교차 검증 점수를 확인
#GradientBoosting은 Decision Tree의 개수를 늘려도 과대적합에 매우 강함

from sklearn.model_selection import cross_validate
from sklearn.ensemble import GradientBoostingClassifier
gb = GradientBoostingClassifier(random_state=42)
scores = cross_validate(gb, train_input, train_target, return_train_score=True, n_jobs=-1)
print(np.mean(scores['train_score']), np.mean(scores['test_score']))


0.8881086892152563 0.8720430147331015


In [6]:
#학습률을 증가시키고 트리의 개수를 늘리면 성능이 향상됨
gb = GradientBoostingClassifier(n_estimators=500, learning_rate = 0.2, random_state=42)
scores = cross_validate(gb, train_input, train_target, return_train_score=True, n_jobs=-1)
print(np.mean(scores['train_score']), np.mean(scores['test_score']))

0.9464595437171814 0.8780082549788999


결정 트리 의 개수를 500개로 (n_estiamtor=500) 늘렸지만, 과대적합을 잘 억제함
학습률 learning_rate의 기본값은 0.1임

In [7]:
gb.fit(train_input, train_target)
print(gb.feature_importances_)

[0.15872278 0.68010884 0.16116839]


각각 [알코올 도수, 당도, pH] (위)

Gradient boossting ([0.15872278 0.68010884 0.16116839])이

Random Forest 보다 ([0.23167441 0.50039841 0.26792718]) 일부 특성 (당도)에 더 집중함

Subsample
- 트리훈련에 사용할 훈련 세트의 비율을 말함
- 기본값 : 1.0 => 전체 훈련 세트 사용, subsample이 1보다 작으면 훈련세트의  일부를 사용 (경사 하강법의 단계마다 일부 sample을 무작위로 선택하여 진행하는 화귤ㄹ적 경사 하강법이나, 미니배치 경사 하강법과 비슷)


일반적으로 Gradient boosting이 random Forest보다 조금 더 높은 성능을 얻을 수 있음

but, gradient boosting이 순서대로 트리를 추가하여 속도가 더 느림.

 즉, GradientBoostingClassifier의  n_jobs매개변수가 없음

